### ISSUE TITLE
libinjection XSS FP with 'on' in URLs

### ISSUE THREAD

**Issue by lifeforms** [SpiderLabs/owasp-modsecurity-crs#663]
Various Referer headers give FP in CRS3. This is extremely annoying, because referers can turn up everywhere and at various URIs. The FP can also happen in other variables, but apparently that is a lot more rare.

Some simplified versions of different examples:

### Example 1: `curl --referer 'http://www.bing.com/search?q=design+ontwerp&go=Search' http://localhost/`

`Message: Warning. detected XSS using libinjection. [file "/usr/local/etc/apache24/security2/crs/rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf"] [line "56"] [id "941100"] [rev "2"] [msg "XSS Attack Detected via libinjection"] [data "Matched Data: referer found within REQUEST_HEADERS:Referer: http://www.bing.com/search?q=design ontwerp&go=Search"] [severity "CRITICAL"] [ver "OWASP_CRS/3.0.0"] [maturity "1"] [accuracy "9"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-xss"] [tag "OWASP_CRS/WEB_ATTACK/XSS"] [tag "WASCTC/WASC-8"] [tag "WASCTC/WASC-22"] [tag "OWASP_TOP_10/A3"] [tag "OWASP_AppSensor/IE1"] [tag "CAPEC-242"]`

### Example 2: `curl --referer 'http://example.com/vacatures?q=Industrieel+Ontwerper&matchtype=b' http://localhost/`

`Message: Warning. detected XSS using libinjection. [file "/usr/local/etc/apache24/security2/crs/rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf"] [line "56"] [id "941100"] [rev "2"] [msg "XSS Attack Detected via libinjection"] [data "Matched Data: referer found within REQUEST_HEADERS:Referer: http://example.com/vacatures?q=Industrieel Ontwerper&matchtype=b"] [severity "CRITICAL"] [ver "OWASP_CRS/3.0.0"] [maturity "1"] [accuracy "9"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-xss"] [tag "OWASP_CRS/WEB_ATTACK/XSS"] [tag "WASCTC/WASC-8"] [tag "WASCTC/WASC-22"] [tag "OWASP_TOP_10/A3"] [tag "OWASP_AppSensor/IE1"] [tag "CAPEC-242"]`

### Example 3: `curl --referer 'http://example.com/search&name=music+on&color=&logo=' http://localhost/`

`Message: Warning. detected XSS using libinjection. [file "/usr/local/etc/apache24/security2/crs/rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf"] [line "56"] [id "941100"] [rev "2"] [msg "XSS Attack Detected via libinjection"] [data "Matched Data: referer found within REQUEST_HEADERS:Referer: http://example.com/search&name=music on&color=&logo="] [severity "CRITICAL"] [ver "OWASP_CRS/3.0.0"] [maturity "1"] [accuracy "9"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-xss"] [tag "OWASP_CRS/WEB_ATTACK/XSS"] [tag "WASCTC/WASC-8"] [tag "WASCTC/WASC-22"] [tag "OWASP_TOP_10/A3"] [tag "OWASP_AppSensor/IE1"] [tag "CAPEC-242"]`

A common pattern in the FPs above is the use of `on` phrase in the query parameters of the referrer URL. E.g. `ontwerp`, `ontwerpen`, `music on`. If you modify the `on` to `ox`, the FP disappears.

So, libinjection sees ` onfoo&bar=` as XSS. This might not be an unreasonable guess, in case there are browsers which "autocorrect" the `&foo` and ignore it. At the other hand, maybe it's overzealous.

What we are sending to libinjection is the URL with `+` converted to the space:

```
[9] T (0) Utf8toUnicode: "http://www.bing.com/search?q=design+ontwerp&go=Search"
[9] T (0) urlDecodeUni: "http://www.bing.com/search?q=design ontwerp&go=Search"
[9] T (0) htmlEntityDecode: "http://www.bing.com/search?q=design ontwerp&go=Search"
[9] T (0) jsDecode: "http://www.bing.com/search?q=design ontwerp&go=Search"
[9] T (0) cssDecode: "http://www.bing.com/search?q=design ontwerp&go=Search"
[9] T (0) removeNulls: "http://www.bing.com/search?q=design ontwerp&go=Search"
[4] Transformation completed in 50 usec.
[4] Executing operator "detectXSS" with param "" against REQUEST_HEADERS:Referer.
[9] Target value: "http://www.bing.com/search?q=design ontwerp&go=Search"
[9] IS_XSS: libinjection detected XSS.
[4] Operator completed in 10 usec.
```

I've checked with a new libinjection from master and it is still firing on these contents:

Payload | Result
--------|-------
`http://www.bing.com/search?q=design ontwerp&go=Search` | xss
`http://example.com/vacatures?q=Industrieel Ontwerper&matchtype=b` | xss
`http://example.com/search&name=music on&color=&logo=` | xss
`onfoo=1` | xss
`onfoo&bar=1` | xss
`fonfoo&bar=1` | not xss
`onfoo&bar&bor=1` | xss

I will create an issue with libinjection, but even if the issue is prevented in upstream, users wil still get hard-to-exclude FPs until they upgrade to ModSecurity 2.9.2.

Considering that, we might consider taking action ourselves such as:

- excluding the Referer header from the libinjection XSS check (for now?)
- try to change transformations, so that `+` is not changed to the space character?

Other possible problems with the rule:

- the rule does not have multimatch like its SQLi brother 942100, was this intentional?
- the 'matched data' in the logdata looks like garbage, libinjection's XSS check only outputs true/false

**Comment by dune73** [comment]
We're between a hammer and a hard place here. The Referer header is part of the standard logformats, it's printed in every log dashboard. It is thus a perfect vector for Log Injections. If we remove libinjection coverage from the referer header, we are opening the door to wide in my opinion.

How frequent is this, and how about moving the Referer header check to PL2 as a stricter sibling?

**Comment by franbuehler** [comment]
I will work on this issue.
I have to take a look at @lifeforms four bullet points to understand them exactly.
From what I understand at the moment, the only thing we can directly control is the first point:
Excluding the referer header from libinjection check (on PL 1 I guess).

**Comment by dune73** [comment]
#717 solving this issue by removing the referer check from PL1 and adding it in a separate rule in PL2 has been merged. Closing this.

**Comment by edgreenberg** [comment]
In what branch would I find this fix?

**Comment by dune73** [comment]
We develop against v3.0/dev at the moment. This will be released as 3.0.1 soon.
